### ISSUE TITLE
Unable to set recordProcessingDelegate when accepting share

### ISSUE THREAD

**Issue by torfluor** [mentrena/SyncKit#155]
When accepting a shared record there is no way, as far as I can see, to set the recordProcessingDelegate of the ModelAdapter that is responsible for the shared record(s). This results in the initial sync of the shared record(s) being run without the delegate and processing of assets is not done.

**Comment by mentrena** [comment]
You're right, it's not an easy use case at the moment.
You could do it in a couple of ways:

- Subclass `DefaultRealmProvider` and override `public func cloudKitSynchronizer(_ synchronizer: CloudKitSynchronizer, modelAdapterForRecordZoneID recordZoneID: CKRecordZone.ID) -> ModelAdapter?` to set the `recordProcessingDelegate`
- Then create your shared synchronizer following the same logic from [here](https://github.com/mentrena/SyncKit/blob/e7f8b477f06140c9b1ab77ef455827f4a05963c7/SyncKit/Classes/RealmSwift/QSCloudKitSynchronizer%2BRealmSwift.swift#L49), but use your subclass as the adapter provider.

Or

- Listen for `Notification.Name.realmProviderDidAddAdapter`
- Then in the handler immediately do something like: 
```swift
guard let zoneId = notification.userInfo["CKRecordZoneID"] else { return } 
synchronizer.adapterProvider.adapterDictionary[zoneId]?.recordProcessingDelegate = yourDelegate
```

I suppose this could be streamlined by adding a delegate to `DefaultRealmProvider`, and it's an easy change so I'll try to add it soon. In the meantime hope this helps.

**Comment by torfluor** [comment]
Listening for realmProviderDidAddAdapter and setting the delegate as you describe worked great!
The adapterDictionary did have internal protection level, so I had to change it to public to get this to work, but that's totally fine for a workaround. 

Thank you for your great support and making this amazing library available!

**Comment by sskjames** [comment]
Hi guys, I stumbled across this problem as well and does this mean, we need to make changes in a fork as DefaultRealmProvider has only `public` access?

Hi @torfluor, did you make those changes in your fork?

**Comment by torfluor** [comment]
Hi @sskjames. Yes, I made adapterDictionary in DefaultRealmProvider.swift public in my locally copy of SyncKit.

I then added the following observer to set the recordProcessingDelegate:

```
        // Workaround for setting recordProcessing Delegate before new share is synced.
        _ = NotificationCenter.default.addObserver(forName: .realmProviderDidAddAdapter, object: nil, queue: OperationQueue.main) { (notification) in
            guard let zoneId = notification.userInfo?["CKRecordZoneID"] as? CKRecordZone.ID else { return }
            
            if let adapterProvider = self.sharedSynchronizer?.adapterProvider as? DefaultRealmProvider {
                adapterProvider.adapterDictionary[zoneId]?.recordProcessingDelegate = self
            }
        }
```

**Comment by sskjames** [comment]
Hi @torfluor, thank you very much for your response. 
Looking at the latest [commits](https://github.com/mentrena/SyncKit/commit/b1992aca9bca3e7357a62d290cb3f36bffdc5e76) to master, I think @mentrena is working towards adding a delegate. I did give it a try, but the `synchronizer(didAddAdapter...)` function in my delegate was not getting invoked. 
Until then, I'll go ahead with this workaround.

**Comment by mentrena** [comment]
Yes sorry, this was released in 1.2.0. 
I ended up adding delegate methods on the synchroniser as I think it makes the use case easier to configure, but I also made `adapterDictionary` in the provider public and the notifications are still called so your code should be fine as well @torfluor.

I'll close here, feel free to reopen if you have any more thoughts on this.
